---
title: SV-eQTLs summary
output: github_document
---

```{r include=FALSE}
knitr::opts_chunk$set(echo=TRUE, message=FALSE, warning=FALSE, fig.width=10)
```

```{r}
library(rtracklayer)
library(dplyr)
library(ggplot2)
library(gridExtra)
library(knitr)
## list of figures
ggp = list()
```

## Read eQTL results

Matrix-eQTL was ran on different sample sets (*all* samples, *Eur*opeans, *YRI*) and using different gene expression normalization (*nonorm*: no additional normalization, *norm*: standardized normal distribution, *quant*: quantile normalization).

```{r}
load('eqtl-test-results.RData')
names(ll)
names(ll) = paste(
  rep(c('Standardization', 'Quantile normalization'),each=3),
  rep(c('EUR + YRI', 'EUR', 'YRI'), 2), sep=' - ')

## list of all genes tested
genes.tested = scan('eqtl-genes.txt', '', quiet=TRUE)
```

## Gene annotation

```{r}
if(!file.exists('../describe-svs/gencode.v35.annotation.gtf.gz')){
  download.file('ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_35/gencode.v35.annotation.gtf.gz', '../describe-svs/gencode.v35.annotation.gtf.gz')
}

genc = import('../describe-svs/gencode.v35.annotation.gtf.gz')
genc = subset(genc, type=='gene')
genc$gene_id = gsub('\\..*', '', genc$gene_id)
genc = genc %>% as.data.frame %>% mutate(gene=gene_id) %>% select(gene, gene_type, gene_name)
```

## Gene families

The gene families were downloaded from [HGNC](https://biomart.genenames.org/martform/#!/default/HGNC?datasets=hgnc_family_mart).

```{r genefam}
genef = read.table('hgnc-families.txt', sep='\t', header=TRUE, quote='')
genef = genef %>% mutate(gene_name=Approved.symbol, gene_family=Family.name) %>%
  select(gene_name, gene_family) %>%
  merge(genc) %>% select(gene, gene_name, gene_family) %>%
  group_by(gene_family) %>%
  mutate(genes.fam=n(), genes.fam.tested=length(unique(intersect(genes.tested, gene))))
```

It contain `r length(unique(genef$gene_family))` gene families. 
For example, a few of the families with more than 100 genes:

```{r genefam2}
genef %>% select(gene_family, genes.fam, genes.fam.tested) %>% unique %>% as.data.frame %>% 
  filter(genes.fam.tested>100) %>% sample_n(10) %>% kable
```

## QC: p-value distribution and QQ plots

```{r qc}
qqplot_matrixeqtl <- function(me.o){
  pvs.s = tibble(n=me.o$cis$hist.counts,
                 pv.s=me.o$cis$hist.bins[-length(me.o$cis$hist.bins)],
                 pv.e=me.o$cis$hist.bins[-1])
  pvs.s$pv.s.exp = cumsum(pvs.s$n)/sum(pvs.s$n)
  pvs.s %>% filter(n>0) %>% 
    ggplot(aes(x=-log10(pv.s.exp), y=-log10(pv.s))) +
    geom_step() +
    ## geom_point(aes(size=cut(n, c(0,1,2,10,Inf), labels=c(1, 2, '3-10', '>10'))), alpha=.5) +
    theme_bw() +
    scale_size_manual(values=c(1,1.5,2,3), 'cis-eQTLs') + 
    geom_abline(linetype=2) +
    xlab('theoritical null p-value (-log10)') +
    ylab('p-value (-log10)')
}

pvhist_matrixeqtl <- function(me.o){
  pvs.s = tibble(n=me.o$cis$hist.counts,
                 pv.s=me.o$cis$hist.bins[-length(me.o$cis$hist.bins)])
  ## pv < 0.01 were split in multiple bins for the qqplot -> aggregate them
  pvs.h = pvs.s %>% mutate(pv.s=ifelse(pv.s<.01, 0, pv.s)) %>%
    group_by(pv.s) %>% summarize(n=sum(n))
  ## histogram
  ggplot(pvs.h, aes(x=pv.s+.005, y=n)) +
    geom_bar(stat='identity', width=.01) + theme_bw() +
    ylim(0, max(pvs.h$n)) +
    xlab('p-value') + ylab('SV-gene test')
}

ggp$hist.all = pvhist_matrixeqtl(ll[['Standardization - EUR + YRI']])
ggp$hist.all + ggtitle('EUR + YRI')
ggp$qq.all = qqplot_matrixeqtl(ll[['Standardization - EUR + YRI']])
ggp$qq.all + ggtitle('EUR + YRI')

ggp$hist.eur = pvhist_matrixeqtl(ll[['Standardization - EUR']])
ggp$hist.eur + ggtitle('EUR')
ggp$qq.eur = qqplot_matrixeqtl(ll[['Standardization - EUR']])
ggp$qq.eur + ggtitle('EUR')

ggp$hist.yri = pvhist_matrixeqtl(ll[['Standardization - YRI']])
ggp$hist.yri + ggtitle('YRI')
ggp$qq.yri = qqplot_matrixeqtl(ll[['Standardization - YRI']])
ggp$qq.yri + ggtitle('YRI')
```

## Number of eQTLs and eGenes

```{r}
eqtl.df = lapply(names(ll)[1:3], function(nn){
  ll[[nn]]$cis$eqtls %>% mutate(exp=nn)
}) %>% bind_rows

eqtl.df = eqtl.df %>% mutate(pop=gsub('.* - (.*)', '\\1', exp),
                             pop=factor(pop, levels=c('EUR + YRI', 'EUR', 'YRI'))) %>%
  select(-exp) %>% dplyr::rename(svid=snps) %>% merge(genc)

svs = read.table('../describe-svs/svs.2504kgp.svsite80al.tsv.gz', as.is=TRUE, header=TRUE)
eqtl.df = svs %>% select(seqnames, start, end, type, size, svid) %>% merge(eqtl.df)

eqtl.df %>% mutate(type='all') %>% rbind(eqtl.df) %>%
  mutate(gene_type=ifelse(gene_type!='protein_coding', 'other', gene_type),
         gene_type=factor(gene_type, levels=c('protein_coding', 'other'))) %>% 
  group_by(gene_type, pop, type) %>%
  summarize(eqtl.fdr01=sum(FDR<=.01),
            esv.fdr01=length(unique(svid[FDR<=.01])),
            egene.fdr01=length(unique(gene[FDR<=.01])),
            eqtl.fdr05=sum(FDR<=.05),
            esv.fdr05=length(unique(svid[FDR<=.05])),
            egene.fdr05=length(unique(gene[FDR<=.05]))) %>%
  kable(format.args=list(big.mark=','))
```

## Positively or negatively correlated with gene expression

```{r effect.size}
eqtl.df %>% filter(FDR<=.01) %>%
  group_by(pop, type) %>% summarize(prop.pos.beta=mean(beta>0),
                                    mean.beta.pos=mean(beta[beta>0]),
                                    mean.beta.neg=mean(beta[beta<0])) %>% kable(digits=3)

eqtl.df %>% filter(FDR<=.01) %>%
  ggplot(aes(beta, fill=type)) +
  geom_histogram(position='dodge') +
  theme_bw() + ylab('SV-eQTL') + 
  facet_grid(pop~., scales='free')
```

More eQTL are associated with an increase in gene expression than a decrease.
The effect size is also stronger.

## Families with eGenes

```{r egene.fam}
## ordered by proportion of family with eQTL
eqtl.fam.df = eqtl.df %>% filter(FDR<=.01) %>% merge(genef) %>%
  group_by(pop) %>% 
  mutate(ntot.egenes=length(unique(gene_name))) %>% 
  group_by(pop, gene_family) %>% summarize(egenes=length(unique(gene_name)),
                                           esvs=length(unique(svid)),
                                           prop=egenes/genes.fam[1],
                                           pv.hyper=1-phyper(q=egenes, m=genes.fam.tested[1],
                                                           n=length(genes.tested)-genes.fam.tested[1],
                                                           k=ntot.egenes[1]),
                                           .groups='drop'
                                           ) %>%
  filter(egenes>1) %>% 
  group_by(pop) %>% mutate(qv.hyper=p.adjust(pv.hyper, method='BH'))

eqtl.fam.df %>% group_by(pop) %>% summarize(qv01=sum(qv.hyper<.01), qv05=sum(qv.hyper<.05))

## top families enriched in the EUR+YRI analysis
eqtl.fam.df %>% filter(qv.hyper<.01, pop=='EUR + YRI', egenes>2, prop>.1) %>%
  arrange(qv.hyper, desc(prop)) %>% ungroup %>% select(-pop) %>% kable
```

Mostly families around immunity (histocompatibility complex, immunoglobulin) and zinc fingers. 
Also anoctamins, mucins, tetraspanins, apolipoproteins.

```{r fam.ex}
eqtl.df %>% filter(FDR<=.01, pop=='EUR + YRI') %>% merge(genef) %>%
  filter(gene_family %in% c('Anoctamins', 'Tetraspanins', 'Apolipoproteins', 'Mucins', 'Aldehyde dehydrogenases')) %>%
  select(gene_family, gene_name, seqnames, start, type, size, beta) %>%
  arrange(gene_family, seqnames, gene_name) %>% 
  kable
```

## SVs associated with multiple genes

```{r multi.genes.qtl}
eqtl.df %>% filter(FDR<=.01, gene_type=='protein_coding') %>% 
  group_by(pop, svid) %>%
  mutate(egenes=length(unique(gene_name))) %>%
  filter(egenes>1) %>%
  group_by(pop) %>% summarize(svs=length(unique(svid)), genes=length(unique(gene_name)))

eqtl.df %>% filter(FDR<=.01, gene_type=='protein_coding') %>% 
  group_by(pop, svid) %>%
  mutate(egenes=length(unique(gene_name))) %>%
  filter(pop=='EUR + YRI', egenes>1, !any(grepl('HLA', gene_name))) %>%
  arrange(desc(egenes), svid) %>%
  ungroup %>%
    mutate(svid=paste0('[', svid,
                        '](https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=',
                        seqnames, '%3A', start, '%2D', end, ')')) %>% 
    select(egenes, svid, type, size, gene_name, beta) %>% 
    head(20) %>% kable
```

### Specific to EUR or YRI

```{r spec}
## frequencies in different super-populations for each SV site
freq.df = read.table('../describe-svs/2504kgp.svsite80al.superpopfreq.tsv.gz', as.is=TRUE, header=TRUE)
freq.df = freq.df %>% group_by(svsite) %>% mutate(af.med=median(af))

pop.spec = freq.df %>%
  filter(abs(af.med-af)>.1, Superpopulation %in% c('AFR', 'EUR'))

## eqtl in eur but not yri or eur+yri
eqtl.eur = eqtl.df %>% group_by(svid, gene) %>%
  filter(n()==1, pop=='EUR', FDR<=.01)

## ex: eqtl in yri but not eur or eur+yri
eqtl.yri = eqtl.df %>% group_by(svid, gene) %>%
  filter(n()==1, pop=='YRI', FDR<=.01)

rbind(
  eqtl.eur %>% mutate(pop='EUR', pop.af=svid %in% pop.spec$svsite),
  eqtl.yri %>% mutate(pop='YRI', pop.af=svid %in% pop.spec$svsite)) %>%
  merge(genc) %>% 
  mutate(gene_type=ifelse(gene_type!='protein_coding', 'other', gene_type),
         gene_type=factor(gene_type, levels=c('protein_coding', 'other'))) %>% 
  group_by(gene_type, pop) %>%
  summarize(eqtl.fdr01=sum(FDR<=.01),
            esv.fdr01=length(unique(svid[FDR<=.01])),
            esv.fdr01.popaf=length(unique(svid[FDR<=.01 & pop.af])),
            egene.fdr01=length(unique(gene[FDR<=.01]))) %>%
  kable
```

*esv.fdr01.popaf*: number of SVs that are eQTLs (FDR<=0.01) and with specific frequency patterns (in EUR or AFR populations).

## Effect of different normalization on the gene expression

```{r norm}
eqtl.all.df = lapply(names(ll), function(nn){
  print(qqplot_matrixeqtl(ll[[nn]]) + ggtitle(nn))
  print(pvhist_matrixeqtl(ll[[nn]]) + ggtitle(nn))
  return(ll[[nn]]$cis$eqtls %>% mutate(exp=nn) %>% filter(FDR<=.01))
}) %>% bind_rows %>% dplyr::rename(svid=snps)

eqtl.all.df %>% merge(genc) %>%
  mutate(gene_type=ifelse(gene_type!='protein_coding', 'other', gene_type),
         gene_type=factor(gene_type, levels=c('protein_coding', 'other'))) %>% 
  group_by(gene_type, exp) %>%
  summarize(eqtl.fdr01=sum(FDR<=.01),
            esv.fdr01=length(unique(svid[FDR<=.01])),
            egene.fdr01=length(unique(gene[FDR<=.01]))) %>%
  kable(format.args=list(big.mark=','))
```

## Examples

```{r ex}
load('./eqtl-examples.RData')

plotEx <- function(ex){
  df = merge(
    tibble(sample=colnames(ge.ex), ge=ge.ex[ex$gene[1],]),
    tibble(sample=colnames(ac.ex), ac=ac.ex[ex$svid[1],])) %>%
    mutate(pop=ifelse(sample %in% yri.samples, 'YRI', 'EUR'))
  df.n = df %>% group_by(ac, pop) %>% summarize(ge=median(ge), n=n())
  ggplot(df, aes(x=factor(ac), y=ge, group=paste(ac, pop))) +
    geom_boxplot(aes(fill=pop)) +
    geom_label(aes(label=n, fill=pop), data=df.n, position=position_dodge(.75)) + 
    theme_bw() +
    scale_fill_brewer(palette='Set2', name='population') +
    xlab('allele count') +
    ylab('gene expression (RPKM)') +
    ggtitle(paste(ex$gene[1], ex$svid[1], paste0(ex$size[1], 'bp'), ex$type[1]))
}
formatEx <- function(df){
  df %>% select(-pop) %>%
    dplyr::rename(svid=snps) %>% merge(svs) %>% 
    mutate(coord=paste0('[', seqnames, ':', start, '-', end,
                        '](https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=',
                        seqnames, '%3A', start, '%2D', end, ')')) %>% 
    select(coord, svid, type, size, gene, gene_name, beta, pvalue, FDR) %>%
    mutate(pvalue=signif(pvalue, 3), FDR=signif(FDR, 3), beta=signif(beta, 3))
}

## most significant positive association in subset of 100 eQTLs
ex = ex.all %>% merge(genc) %>% filter(statistic>0, gene_type=='protein_coding') %>%
  arrange(FDR) %>% head(1) %>% formatEx
ex %>% select(-gene) %>% kable
ggp$ex.all.pos = plotEx(ex)
ggp$ex.all.pos

## most significant negative association in subset of 100 eQTLs
ex = ex.all %>% merge(genc) %>% filter(statistic<0, gene_type=='protein_coding', !grepl('HLA', gene_name)) %>%
  arrange(FDR) %>% head(1) %>% formatEx
ex %>% select(-gene) %>% kable
ggp$ex.all.neg = plotEx(ex)
ggp$ex.all.neg

## most significant positive association in YRI-specific eQTLs
ex = ex.yri %>% merge(genc) %>% filter(statistic>0, gene_type=='protein_coding') %>%
  arrange(FDR) %>% head(1) %>% formatEx
ex %>% select(-gene) %>% kable
ggp$ex.yri.pos = plotEx(ex)
ggp$ex.yri.pos

## most significant negative association in YRI-specific eQTLs
ex = ex.yri %>% merge(genc) %>% filter(statistic<0, gene_type=='protein_coding') %>%
  arrange(FDR) %>% head(1) %>% formatEx
ex %>% select(-gene) %>% kable
ggp$ex.yri.neg = plotEx(ex)
ggp$ex.yri.neg
```

## Multi-panel figure

```{r fig}
## adds a legend title: a), b), etc
plot_list <- function(ggp.l, gg.names=NULL, gg.titles=NULL){
  if(is.null(names(ggp.l))) names(ggp.l) = paste0('g', 1:length(ggp.l))
  if(is.null(gg.names)) gg.names = names(ggp.l)
  if(is.null(gg.titles)){
    gg.titles = rep('', length(gg.names))
  } else {
    gg.titles = paste0(' ', gg.titles)
  }
  lapply(1:length(gg.names), function(ii) ggp.l[[gg.names[ii]]] + ggtitle(paste0('(', LETTERS[ii], ')', gg.titles[ii])))
}

grid.arrange(grobs=plot_list(ggp,
                             gg.names=names(ggp)[1:6],
                             gg.titles=rep(c('EUR+YRI', 'EUR', 'YRI'), each=2)),
             layout_matrix=matrix(1:6, 2),
             heights=c(3, 5))

## grid.arrange(grobs=plot_list(ggp,
##                              gg.names=names(ggp)[7:10],
##                              gg.titles=rep(c('EUR+YRI', 'YRI'), each=2)),
##              layout_matrix=matrix(1:4, 2))

pdf('fig-sv-eqtl.pdf', 9, 5)
grid.arrange(grobs=plot_list(ggp,
                             gg.names=names(ggp)[1:6],
                             gg.titles=rep(c('EUR+YRI', 'EUR', 'YRI'), each=2)),
             layout_matrix=matrix(1:6, 2),
             heights=c(3, 5))
dev.off()
```

## Save table

```{r tab}
tab = eqtl.df %>% mutate(type='all') %>% rbind(eqtl.df) %>%
  mutate(gene_type=ifelse(gene_type!='protein_coding', 'other', gene_type),
         gene_type=factor(gene_type, levels=c('protein_coding', 'other'))) %>% 
  group_by(gene_type, pop, type) %>%
  summarize(eQTL=sum(FDR<=.01),
            eSV=length(unique(svid[FDR<=.01])),
            eGene=length(unique(gene[FDR<=.01])))

kable(tab, format.args=list(big.mark=','))
kable(tab, format.args=list(big.mark=','), format='latex') %>% cat(file='eqtl-summary.tex')
```

## Save eQTL information

```{r save}
eqtl.df %>% filter(FDR<=.01) %>% 
  select(seqnames, start, end, svid, type, size, pop,
         gene, gene_name, gene_type, beta, pvalue, FDR) %>%
  arrange(FDR) %>% 
  write.table(file='vggiraffe-sv-eqtl-geuvadis.FDR01.csv', sep=',', quote=FALSE, row.names=FALSE)

## enriched families
eqtl.fam.df %>% filter(qv.hyper<=.01, pop=='EUR + YRI', egenes>1) %>%
  arrange(qv.hyper, desc(prop)) %>% ungroup %>% select(-pop) %>%
    write.table(file='vggiraffe-geuvadis-sveqtl-gene-families.csv', sep=',', quote=TRUE, row.names=FALSE)
```
